File: src/main/java/xyz/oribuin/skyblock/command/argument/MemberArgument.java
package xyz.oribuin.skyblock.command.argument;

import dev.rosewood.rosegarden.command.framework.Argument;
import dev.rosewood.rosegarden.command.framework.ArgumentHandler;
import dev.rosewood.rosegarden.command.framework.CommandContext;
import dev.rosewood.rosegarden.command.framework.InputIterator;
import xyz.oribuin.skyblock.island.member.Member;

import java.util.List;

public class MemberArgument extends ArgumentHandler<Member> {

    public MemberArgument() {
        super(Member.class);
    }

    @Override
    public Member handle(CommandContext context, Argument argument, InputIterator inputIterator) throws HandledArgumentException {
        return null;
    }

    @Override
    public List<String> suggest(CommandContext context, Argument argument, String[] args) {
        return null;
    }

}

File: Plugin/src/main/kotlin/xyz/oribuin/skyblock/command/argument/MemberArgument.kt
package xyz.oribuin.skyblock.command.argument

import dev.rosewood.rosegarden.RosePlugin
import dev.rosewood.rosegarden.command.framework.ArgumentParser
import dev.rosewood.rosegarden.command.framework.RoseCommandArgumentHandler
import dev.rosewood.rosegarden.command.framework.RoseCommandArgumentInfo
import dev.rosewood.rosegarden.utils.StringPlaceholders
import org.bukkit.Bukkit
import xyz.oribuin.skyblock.island.Member
import xyz.oribuin.skyblock.manager.IslandManager
import xyz.oribuin.skyblock.util.getManager

class MemberArgument(rosePlugin: RosePlugin) : RoseCommandArgumentHandler<Member>(rosePlugin, Member::class.java) {

    private val islandManager = this.rosePlugin.getManager<IslandManager>()

    override fun handleInternal(argumentInfo: RoseCommandArgumentInfo, argumentParser: ArgumentParser): Member {
        val input = argumentParser.next()

        val exception = HandledArgumentException("argument-handler-member-option", StringPlaceholders.single("input", input))

        return this.islandManager.getMember(Bukkit.getOfflinePlayerIfCached(input) ?: throw exception)
    }

    override fun suggestInternal(argumentInfo: RoseCommandArgumentInfo, argumentParser: ArgumentParser): MutableList<String> {
        argumentParser.next()
        return Bukkit.getOfflinePlayers().mapNotNull { it.name }.toMutableList()
    }

}